Allow displaying videos without transcripts

OmarIthawi opened the pull request:
## Description
This is a bug for mobile we got it reported and fixed earlier in our fork by @ahmedaljazzar:
 - https://github.com/Edraak/edx-platform/pull/279
 - https://github.com/Edraak/edx-platform/issues/290

## Background
As far as I know, at edX every video has to have transcripts. So it's safe to assume that English transcripts is always a working fallback. The accessibility is something that is well invested in the US and the English language in general, especially with the aid of YouTube's free transcripters and the large community of volunteers.

However, in our case @Edraak it's a different story. We can't always afford having transcripts for our videos. So, that's a broken assumption. Therefore this PR is created with backward compatibility in mind.

## TODO
 - [x] Figure out how to properly make an `xmodule` test work with `FEATURES`?
 - [x] Unit tests and quality checks
 - [x] Get engineering/product feedback from @edX
 - [x] Get additional review from @ahmedaljazzar 
 - [x] Should `'en'` be changed to `settings.LANGUAGE_CODE` instead? (looks like it!)
 - [x] Manual testing with mobile app
   (or the API directly `/api/mobile/v0.5/video_outlines/courses/course-v1:Edraak+JAZ101+2017`)

## Sandbox
- [x] https://default-english-transcripts.sandbox.edx.org/

## Bug
Video player: Caption icon should not be displayed for the video doesn't have a transcript.

### Before
Notice that the "English" transcript option is there, but the video has no transcripts at all:
<img width="250" src="https://user-images.githubusercontent.com/645156/27902004-bfb38944-623c-11e7-828a-c1a89fc1684f.png" />


### After
Notice the "English" options is no longer showing:
<img width="250" src="https://user-images.githubusercontent.com/645156/27902016-c8994ddc-623c-11e7-8398-543ba49dfd82.png" />

edx-webhook commented:
Thanks for the pull request, @OmarIthawi! I've created [OSPR-1811](https://openedx.atlassian.net/browse/OSPR-1811) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

OmarIthawi commented:
@ahmedaljazzar عطوفة السيد 

يرجى الاطلاع مع الاحترام

OmarIthawi commented:
Ping @gsong 😄

OmarIthawi commented:
Just a gentle reminder @gsong 😃

OmarIthawi commented:
jenkins run all

ahmedaljazzar commented on common/lib/xmodule/xmodule/video_module/video_handlers.py:
```diff
@@ -276,7 +276,7 @@ def transcript(self, request, dispatch):
         elif dispatch.startswith('available_translations'):
 
-            available_translations = self.available_translations(transcripts)
+            available_translations = self.available_translations(transcripts, verify_assets=True)
```
Why aren't we falling back to English transcripts here?

Qubad786 replied:
I don't see any good reason for not having consistency among the usages of `verify_assets`.

_Lets assume_ `FALLBACK_TO_ENGLISH_TRANSCRIPTS=True`
On a fresh load, we won't verify assets and language "English" shows up in lang menu
And, selecting "English" from the lang menu hides it away from that menu (i.e. due to an ajax call to this more strict dispatch)

@OmarIthawi, do you see any issues if we totally depend on feature `FALLBACK_TO_ENGLISH_TRANSCRIPTS`? 
I don't think, there is any point of having a language in the menu whose content is not in place (i.e. not verified)

@muzaffaryousaf I'd like to hear your thoughts on this too.

OmarIthawi replied:
@Qubad786, sorry didn't see your comment earlier. `FALLBACK_TO_ENGLISH_TRANSCRIPTS` should incur a small performance penalty. Which is the point of returning `en` always. Not sure what is the real cost of enabling it at the scale of edX.

But at Edraak we're happy with this verification and didn't notice any penalty.

After all, it's just a feature, so the DevOps team can toggle and monitor it.

OmarIthawi commented on common/lib/xmodule/xmodule/video_module/video_handlers.py:
```diff
@@ -276,7 +276,7 @@ def transcript(self, request, dispatch):
         elif dispatch.startswith('available_translations'):
 
-            available_translations = self.available_translations(transcripts)
+            available_translations = self.available_translations(transcripts, verify_assets=True)
```
This used to be `verify=True` implicitly, now it's explicit since the signature of `available_translations()` have changed.

ahmedaljazzar replied:
So what's the idea of falling back to english if we still forcing some modules to verify/not-verify the assets?

OmarIthawi replied:
In this specific line, I'm being totally backward compatible. So, it used to verify in `master` (so to speak), and I'd like my PR to do the same here.

The other option, where `master` used to skip verification, this PR allows enforcing verifications in those cases.

Old Logic -> New Logic:
 - Verified (by default with function signature) -> Verified (enforced by the callee) = Logic haven't changed, backward compatible and hard-coded.
 - Not verified (by callee) -> Configurable (by feature flag)

I hope that explains it

OmarIthawi commented:
@gsong Could you please invite someone to take a look here (perhaps @BenjiLee)?

gsong commented:
@edx/learner 🛎

muzaffaryousaf commented:
@gsong @OmarIthawi I think my team can take a look at this since it's related to videos but it won't be this week at-least.

catong commented:
@srpearce FYI

srpearce commented:
So... For this PR, are we documenting how to prevent transcript options from showing if a video doesn't have transcripts? Also, this will apply only to Open edX, correct?

gsong commented:
@srpearce If truly all edx.org and edge content has transcriptions, then this would only apply to Open edX.

OmarIthawi commented:
> So... For this PR, are we documenting how to prevent transcript options from showing if a video doesn't have transcripts? Also, this will apply only to Open edX, correct?

That's correct @srpearce (As far as I know!)

marcotuts commented:
@gsong  - Components inside of course content (video, html, capa, etc) we generally group with Educator. I'm going to flag @sstack22 here as an FYI (from a product standpoint). @muzaffaryousaf  is already on the eng review. 👍  

Separately, we can probably help you out by having an easy to reference "platform components by team" document / listing that is complete :) I can raise this within the product group next week.

srpearce commented:
So... how *does* one prevent transcript options from showing if a video doesn't have transcripts? Is this a Studio setting? Is it something the course team does, or the administrator who's running the instance of Open edX?

OmarIthawi commented:
Hey @srpearce, this is a setting that the administrator should configure. It is a setting that reflects the platform-wide content policy: whether it enforces having English transcripts on every video or not.

srpearce commented:
@grantgoodmanedX Since this info is for the ICR guide, do you mind taking over? It looks like this is a single a platform-wide setting that system administrators use to allow videos to not have English transcripts.

grantgoodmanedX commented:
@srpearce I'll be happy to.

srpearce commented:
@grantgoodmanedX Great, thanks!

Qubad786 commented:
Hey @OmarIthawi, would you please rebase this PR with edx:master? I'm having trouble building sandbox for your branch. 
traceback:
```
08:06:00 failed: [default-english-transcripts.sandbox.edx.org] (item=lms) => {"changed": true, "cmd": 
["/edx/bin/edxapp-migrate-lms"], "delta": "0:00:00.988104", "end": "2017-08-31 08:06:00.757225", 
"failed": true, "item": "lms", "rc": 1, "start": "2017-08-31 08:05:59.769121", "stderr": "Traceback (most 
recent call last):\n  File \"manage.py\", line 111, in <module>\n    startup = 
importlib.import_module(edx_args.startup)\n  File \"/usr/lib/python2.7/importlib/__init__.py\", line 37, in 
import_module\n    __import__(name)\n  File \"/edx/app/edxapp/edx-platform/lms/startup.py\", line 12, 
in <module>\n    settings.INSTALLED_APPS  # pylint: disable=pointless-statement\n  File 
\"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/conf/__init__.py\", line 48, in 
__getattr__\n    self._setup(name)\n  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-
packages/django/conf/__init__.py\", line 44, in _setup\n    self._wrapped = Settings(settings_module)\n  
File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/conf/__init__.py\", line 
92, in __init__\n    mod = importlib.import_module(self.SETTINGS_MODULE)\n  File 
\"/usr/lib/python2.7/importlib/__init__.py\", line 37, in import_module\n    __import__(name)\n  File 
\"/edx/app/edxapp/edx-platform/lms/envs/aws.py\", line 215, in <module>\n    BOOK_URL = 
ENV_TOKENS['BOOK_URL']\nKeyError: 'BOOK_URL'", "stdout": "", "stdout_lines": [], "warnings": []} 
```

Thanks!

Qubad786 commented:
Sandbox is ready for this PR.
https://default-english-transcripts.sandbox.edx.org/

OmarIthawi commented:
jenkins run bokchoy

OmarIthawi commented:
jenkins run coverage

OmarIthawi commented:
Hey @Qubad786, the PR should be ready for testing. I assume that `grantgoodmanedX` is going to test this, right?

muzaffaryousaf commented:
@grantgoodmanedX Any update on this PR from doc perspective, we've sandbox ready and will be able to merge probably tomorrow.

grantgoodmanedX commented:
@OmarIthawi I won't be testing this; I do plan to add information about it to the Open edX installation guide.
@muzaffaryousaf You don't need to wait for my documentation changes to be ready before you merge it. But I should have a draft ready for review tomorrow.

OmarIthawi commented:
@Qubad786 would you mind not merging this for now? @ahmedaljazzar is testing this on a mobile app. I guess one final check won't hurt.

ahmedaljazzar commented:
Hello @Qubad786, I tested the feature on the [sandbox](https://default-english-transcripts.sandbox.edx.org/) and the current edX behavior works perfect. Could we redeploy it with a feature flag `FALLBACK_TO_ENGLISH_TRANSCRIPTS ` set to `False` to test the other behavior?

Qubad786 commented:
@ahmedaljazzar no need to redeploy, I've toggled it. You should be able to test it.

ahmedaljazzar commented:
@Qubad786, did you restart the server after toggling the feature flag? Otherwise @OmarIthawi we might have an issue in the PR because I still can see the endpoint returning the `en` transcript for all videos that doesn't have a transcript.

OmarIthawi commented:
@ahmedaljazzar the feature is on both LMS/CMS and requires a server restart like you mentioned.

Qubad786 commented:
@ahmedaljazzar yes, I've also verified `FALLBACK_TO_ENGLISH_TRANSCRIPTS`, its `False`.
```
from django.conf import settings; settings.FEATURES['FALLBACK_TO_ENGLISH_TRANSCRIPTS']
```

OmarIthawi commented:
@Qubad786 and @ahmedaljazzar Let me debug it further to double check.

OmarIthawi commented:
@Qubad786 and @ahmedaljazzar, I've just finished debugging it and fixed the bug. I've also added one integration test in `mobile_api/video_outlines/tests.py` to make sure it works!

OmarIthawi commented:
jenkins run bokchoy

Qubad786 commented:
jenkins run bokchoy

Qubad786 commented:
@ahmedaljazzar, @OmarIthawi here is the updated sandbox: https://default-english-transcripts.sandbox.edx.org

ahmedaljazzar commented:
Thanks @Qubad786, the default behavior works perfect, could you please toggle the flag?

Qubad786 commented:
@ahmedaljazzar I've toggled it.

ahmedaljazzar commented:
It worked perfect now. I tested the feature for an empty and not-empty transcripts for its both values and it behaves as expected.

OmarIthawi commented:
Hey @Qubad786, just a reminder about this.

grantgoodmanedX commented:
@OmarIthawi For documentation purposes, I want to understand the default behavior. Is this new default behavior, so that if there is no transcript at all, the transcript button doesn't display? Or do users who want this behavior need to set some configuration, like setting FALLBACK_TO_ENGLISH_TRANSCRIPTS to FALSE?

OmarIthawi commented:
@grantgoodmanedX 

> Is this new default behavior, so that if there is no transcript at all, the transcript button doesn't display?

No, it's the other way around. Since it's an edX requirement (from both product management and performance perspective).

> users who want this behavior need to set some configuration, like setting FALLBACK_TO_ENGLISH_TRANSCRIPTS to FALSE?

Yes, this is the correct behavior.

OmarIthawi commented:
Thanks @Qubad786. I've used the `assertEqual()` function.

grantgoodmanedX commented:
I've created a PR in the edx-documentation repo https://github.com/edx/edx-documentation/pull/1589/ . Please take a look when you have a chance. I'm not sure I understand how the configuration worked and assumed that it's like other feature flags.

OmarIthawi commented:
🎉 

Thanks @Qubad786 and @grantgoodmanedX!

Qubad786 commented:
@OmarIthawi Thank you for your contribution. 👏

edx-webhook commented:
@OmarIthawi 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/15475&entry.1484655318=https://github.com/OmarIthawi&entry.752974735=2017-07-06+08:24&entry.1917517419=2017-09-12+07:22&entry.2133058324=Yes) so we can improve your experience in the future.
